code/go/internal/validator/spec.go

```diff
@@ -102,36 +103,77 @@ func substringInSlice(str string, list []string) bool {
+		matcher *regexp.Regexp
+		new     string
 	}{
 		{
-			original: "Must not validate the schema (not)",
-			new:      "Must not be present",
+			matcher: regexp.MustCompile(`Must not validate the schema \(not\)`),
+			new:     "Must not be present",
+		},
+		{
+			matcher: regexp.MustCompile("secret is required"),
+			new:     "variable identified as possible secret, secret parameter required to be set to true or false",
+		},
+		{
+			matcher: regexp.MustCompile(`(field processors.[0-9]+.rename): if is required`),
+			new:     "%s: rename \"message\" to \"event.original\" processor requires if: 'ctx.event?.original == null'",
 		},
 		{
-			original: "secret is required",
-			new:      "variable identified as possible secret, secret parameter required to be set to true or false",
+			matcher: regexp.MustCompile(`(field processors.[0-9]+): remove is required`),
+			new:     "%s: rename \"message\" to \"event.original\" processor requires remove \"message\" processor",
+		},
+		{
+			matcher: regexp.MustCompile(`(processors.[0-9]+.remove.field): processors.[0-9]+.remove.field does not match: "message"`),
+			new:     "%s: rename \"message\" to \"event.original\" processor requires remove \"message\" processor",
+		},
+		{
+			matcher: regexp.MustCompile(`(processors.[0-9]+.remove.if): processors.[0-9]+.remove.if does not match: "ctx\.event\?\.original != null"`),
+			new:     "%s: rename \"message\" to \"event.original\" processor requires remove \"message\" processor with if: 'ctx.event?.original != null'",
 		},
 	}
+
+	// Filter out redundant errors
 	redundant := []string{
 		"Must validate \"then\" as \"if\" was valid",
 		"Must validate \"else\" as \"if\" was not valid",
 		"Must validate all the schemas (allOf)",
 		"Must validate at least one schema (anyOf)",
 		"Must validate one and only one schema (oneOf)",
+		"At least one of the items must match",
 	}
+
+	// Add error code to specific errors
+	addErrorCode := []struct {
+		matcher *regexp.Regexp
+		code    string
+	}{
+		{
+			matcher: regexp.MustCompile(`rename "message" to "event.original" processor`),
+			code:    specerrors.MessageRenameToEventOriginalValidation,
+		},
+	}
+
 	for _, e := range errs {
 		for _, msg := range msgTransforms {
-			if strings.Contains(e.Error(), msg.original) {
+			if match := msg.matcher.FindStringSubmatch(e.Error()); len(match) > 1 {
+				e = specerrors.NewStructuredError(
+					errors.New(strings.Replace(e.Error(), match[0], fmt.Sprintf(msg.new, match[1]), 1)),
+					specerrors.UnassignedCode)
+			} else if msg.matcher.MatchString(e.Error()) {
 				e = specerrors.NewStructuredError(
-					errors.New(strings.Replace(e.Error(), msg.original, msg.new, 1)),
+					errors.New(strings.Replace(e.Error(), msg.matcher.FindString(e.Error()), msg.new, 1)),
 					specerrors.UnassignedCode)
 			}
 		}
+		for _, transform := range addErrorCode {
+			if transform.matcher.MatchString(e.Error()) {
+				e = specerrors.NewStructuredError(e, transform.code)
```

Comment:
Should we create here an new error to avoid wrapping a StructuredError inside another structuredError ? 

StructuredError implements `UnWrap`, so not sure if this is a problem or not @jsoriano WDYT ?


If that could be a problem, this could be changed by something like:

```suggestion
				e = specerrors.NewStructuredError(error.New(e.Error()), transform.code)
```

Reply:
TLDR; Any option looks good to me for this PR, but let's keep the discussion for a follow up.

I see that in other transforms we are creating new errors. For consistency we should do the same thing in all cases.

Regarding wrapping structured errors inside structured errors, I don't see why this should be a problem, maybe we have to clarify the semantics in the `specerrors` package.

It would be probably better to wrap the error as is instead of creating new ones. By creating new ones we are losing information. For example an structured error could be wrapping an `os.ErrNotFound` error, and we would be losing this information with `error.New(e.Error())`.

So any option looks good to me for this PR, but we should follow up clarifying these cases. If wrapping structured errors is a problem, `NewStructuredError` could take care of this, perhaps with something like this:
```
// NewStructuredError creates a generic validation error by wrapping the given error. If the given error
// is already a StructuredError, the returned error wraps the error wrapped by the given error.
func NewStructuredError(err error, code string) *StructuredError {
        if se, ok := err.(*StructuredError) {
                result := *se
                result.code = code
                return &result
        }
        return &StructuredError{
                err:  err,
                code: code,
        }
}
```